Prevent unescaped query from being returned to clients

acharis opened the pull request:
The context: https://vitess.slack.com/archives/C0PQY0PTK/p1520444718000317

I'm stripping out the query from the SQLError, because we're going to add in a properly escaped version right after. I've tried to leave the terseErrors codepath alone.

Signed-off-by: Alex Charis <acharis@hubspot.com>

acharis commented:
@demmer @sougou @bbeaudreault

alainjobart commented on go/vt/vttablet/tabletserver/tabletserver.go:
```diff
@@ -2050,12 +2053,15 @@ func (tsv *TabletServer) GetTxPoolWaiterCap() int64 {
 // and also truncates data if it's too long
```
This comment is not correct, there is no truncate.

Also, I would document that bindVariables=nil has a different meaning than len(bindVariables]==0, as it's not common to do so. And the caller depends on this behavior.

alainjobart commented on go/vt/vttablet/tabletserver/tabletserver.go:
```diff
@@ -1402,34 +1390,49 @@ func (tsv *TabletServer) convertAndLogError(ctx context.Context, sql string, bin
+	if maxLen != 0 && len(message) > maxLen {
+		message = message[:maxLen-12] + " [TRUNCATED]"
+	}
+	if logMethod != nil {
```
I would move this up before line 1414, so we don't truncate the message if we're not going to use it.

alainjobart commented on go/vt/vttablet/tabletserver/tabletserver.go:
```diff
@@ -1402,34 +1390,49 @@ func (tsv *TabletServer) convertAndLogError(ctx context.Context, sql string, bin
+	} else {
+		err = vterrors.Errorf(errCode, "%v%s", err.Error(), callerID)
+		if tsv.TerseErrors && len(bindVariables) != 0 && errCode != vtrpcpb.Code_FAILED_PRECONDITION {
+			message = fmt.Sprintf("%v: %v", err, queryAsString(sql, nil))
```
This is a change in behavior, isn't it? We won't log the bind variables in this case, but we were logging them before.

demmer replied:
Alain is correct... the tests show this change in behavior:

```
 --- FAIL: TestTerseErrorsBindVars (0.46s)
	tabletserver_test.go:2624: error log '(errno 10) (sqlstate HY000) during query: select * from test_table where a = :a', want 'sensitive message (errno 10) (sqlstate HY000) during query: select * from test_table where a = 1: Sql: "select * from test_table where a = :a", BindVars: {a: "type:INT64 value:\"1\" "}'
```

The behavior we want is for the full mysql error message and all bind vars to be logged to the local log, but the stripped error to be returned back over the wire.

acharis commented:
Yes, it's a change in behaviour, and to make things more concrete, here are some log+error message snippets from before and after this change, with terse on and off:

OLD:

from log of java service (terse = false):
org.skife.jdbi.v2.exceptions.UnableToExecuteStatementException: java.sql.SQLSyntaxErrorException: legacy_code: BAD_INPUT_LEGACY
message: "vtgate: http://app-y-vtgate-us-east-1a-539718136-g0gz5:15101/: target: vitamins.0.master, used tablet: qa_iad-1350593002 (172.18.125.176), vttablet: rpc error: code = InvalidArgument desc = table snerb not found in schema, CallerID: acharis.hs-dataops.vitess"
code: INVALID_ARGUMENT
 [statement:"select id, data from snerb where id=:id", located:"select id, data from snerb where id=:id", rewritten:"select /*+ MAX_EXECUTION_TIME(30000) */ id, data from snerb where id=?", arguments:{ positional:{}, named:{id:59987}, finder:[]}]

tablet log (terse = false):
I0312 21:48:42.341763     104 tabletserver.go:1340] table snerb not found in schema, CallerID: acharis.hs-dataops.vitess: Sql: "select /*+ MAX_EXECUTION_TIME(30000) */ id, data from snerb where id = :v1", BindVars: {v1: "type:INT64 value:\"59987\" "}

from log of java service (terse = true):
org.skife.jdbi.v2.exceptions.UnableToExecuteStatementException: java.sql.SQLSyntaxErrorException: legacy_code: BAD_INPUT_LEGACY
message: "vtgate: http://app-y-vtgate-us-east-1e-3317181431-vvpjh:15101/: target: vitamins.0.master, used tablet: qa_iad-1350593001 (172.18.178.204), vttablet: rpc error: code = InvalidArgument desc = table snerb not found in schema, CallerID: acharis.hs-dataops.vitess"
code: INVALID_ARGUMENT
 [statement:"select id, data from snerb where id=:id", located:"select id, data from snerb where id=:id", rewritten:"select /*+ MAX_EXECUTION_TIME(30000) */ id, data from snerb where id=?", arguments:{ positional:{}, named:{id:59987}, finder:[]}]

I0312 22:30:03.519956    1772 tabletserver.go:1340] table snerb not found in schema, CallerID: acharis.hs-dataops.vitess: Sql: "select /*+ MAX_EXECUTION_TIME(30000) */ id, data from snerb where id = :v1", BindVars: {v1: "type:INT64 value:\"59987\" "}

NEW:

from log of java service (terse = false):
org.skife.jdbi.v2.exceptions.UnableToExecuteStatementException: java.sql.SQLSyntaxErrorException: legacy_code: BAD_INPUT_LEGACY
message: "vtgate: http://app-z-vtgate-us-east-1a-4136914187-3vp0f:15101/: target: vitamins.0.master, used tablet: qa_iad-1350593002 (172.18.125.224), vttablet: rpc error: code = InvalidArgument desc = table snerb not found in schema (CallerID: acharis.hs-dataops.vitess)"
code: INVALID_ARGUMENT
 [statement:"select id, data from snerb where id=:id", located:"select id, data from snerb where id=:id", rewritten:"select /*+ MAX_EXECUTION_TIME(30000) */ id, data from snerb where id=?", arguments:{ positional:{}, named:{id:59987}, finder:[]}]

tablet log (terse = false):
I0312 22:45:38.586857     141 tabletserver.go:1407] table snerb not found in schema (CallerID: acharis.hs-dataops.vitess): Sql: "select /*+ MAX_EXECUTION_TIME(30000) */ id, data from snerb where id = :v1", BindVars: {v1: "type:INT64 value:\"59987\" "}

from log of java service (terse = true):
org.skife.jdbi.v2.exceptions.UnableToExecuteStatementException: java.sql.SQLSyntaxErrorException: legacy_code: BAD_INPUT_LEGACY
message: "vtgate: http://app-y-vtgate-us-east-1b-1158393513-mcv1p:15101/: target: vitamins.0.master, used tablet: qa_iad-1350593001 (172.18.189.177), vttablet: rpc error: code = InvalidArgument desc = table snerb not found in schema (CallerID: acharis.hs-dataops.vitess)"
code: INVALID_ARGUMENT
 [statement:"select id, data from snerb where id=:id", located:"select id, data from snerb where id=:id", rewritten:"select /*+ MAX_EXECUTION_TIME(30000) */ id, data from snerb where id=?", arguments:{ positional:{}, named:{id:59987}, finder:[]}]

tablet log (terse = true):
I0312 23:09:18.111559     103 tabletserver.go:1407] table snerb not found in schema (CallerID: acharis.hs-dataops.vitess): Sql: "select /*+ MAX_EXECUTION_TIME(30000) */ id, data from snerb where id = :v1"

acharis commented:
(i've deleted the old comment, which didn't properly show the differences.)
Yes, it's a change in behaviour, and to make things more concrete, here are some error message snippets from before and after this change, with terse set to off:

NEW error msg returned to java client:
```
2018-04-12 14:32:19.952 [dw-103 - GET /v1/vitamins/insert_foo] ERROR c.h.j.f.r.CoreResponseListenerModule - server error 500 handling 'http://localhost:8080/v1/vitamins/insert_foo'
org.skife.jdbi.v2.exceptions.UnableToExecuteStatementException: java.sql.SQLIntegrityConstraintViolationException: legacy_code: INTEGRITY_ERROR_LEGACY
message: "vtgate: http://app-x-vtgate-us-east-1a-2085821795-7h853:15101/: execInsertUnsharded: target: vitamins.0.master, used tablet: qa_iad-1350593001 (172.18.185.177), vttablet: rpc error: code = AlreadyExists desc = Duplicate entry \'1\' for key \'PRIMARY\' (errno 1062) (sqlstate 23000) (CallerID: acharis.hs-dataops.vitess): Sql: \"insert into foo(id, process) values (:v1, :v2)\", BindVars: {v1: \"type:INT64 value:\\\"1\\\" \"v2: \"type:VARBINARY value:\\\"Ot\\\\\\\\Ll\\\\235Ia\\\\211\\\\020\\\\374\\\\373\\\\204w\\\\310\\\\005\\\" \"#maxLimit: \"type:INT64 value:\\\"100001\\\" \"}"
code: ALREADY_EXISTS
 [statement:"INSERT INTO foo (id, process) values (:id, :process)", located:"INSERT INTO foo (id, process) values (:id, :process)", rewritten:"INSERT INTO foo (id, process) values (?, ?)", arguments:{ positional:{}, named:{process:[79, 116, 92, 76, 108, -99, 73, 97, -119, 16, -4, -5, -124, 119, -56, 5],id:1}, finder:[]}]
        at org.skife.jdbi.v2.SQLStatement.internalExecute(SQLStatement.java:1338)
        at org.skife.jdbi.v2.Update.execute(Update.java:56)
        at org.skife.jdbi.v2.sqlobject.UpdateHandler$2.value(UpdateHandler.java:68)
        at org.skife.jdbi.v2.sqlobject.UpdateHandler.invoke(UpdateHandler.java:81)
        at org.skife.jdbi.v2.sqlobject.SqlObject.invoke(SqlObject.java:216)
```

OLD error msg returned to java client:
```
2018-04-12 14:36:57.815 [dw-102 - GET /v1/vitamins/insert_foo_todo] ERROR c.h.j.f.r.CoreResponseListenerModule - server error 500 handling 'http://localhost:8080/v1/vitamins/insert_foo_todo'
org.skife.jdbi.v2.exceptions.UnableToExecuteStatementException: java.sql.SQLNonTransientException: gRPC StatusRuntimeException: hidden.vitess.io.grpc.StatusRuntimeException: CANCELLED: Failed to read message. [statement:"INSERT INTO foo (id, process) values (:id, :process)", located:"INSERT INTO foo (id, process) values (:id, :process)", rewritten:"INSERT INTO foo (id, process) values (?, ?)", arguments:{ positional:{}, named:{process:[-18, -44, -45, 69, 85, 57, 71, -73, -79, -124, -12, 21, 117, -30, 67, 24],id:1}, finder:[]}]
        at org.skife.jdbi.v2.SQLStatement.internalExecute(SQLStatement.java:1338)
        at org.skife.jdbi.v2.Update.execute(Update.java:56)
        at org.skife.jdbi.v2.sqlobject.UpdateHandler$2.value(UpdateHandler.java:68)
        at org.skife.jdbi.v2.sqlobject.UpdateHandler.invoke(UpdateHandler.java:81)
        at org.skife.jdbi.v2.sqlobject.SqlObject.invoke(SqlObject.java:216)

```

this shows exactly what our issue is: with the code previous to this PR, the client doesn't see the mysql error "Duplicate entry" because the query isn't UTF-8.

acharis commented:
For whatever reason, I can't find anything in the tablet log for these failing queries.

acharis commented:
More log message examples:

NEW message returned to java client:
```
org.skife.jdbi.v2.exceptions.UnableToExecuteStatementException: java.sql.SQLSyntaxErrorException: legacy_code: BAD_INPUT_LEGACY
message: "vtgate: http://app-x-vtgate-us-east-1b-2082136550-622tt:15101/: execInsertUnsharded: target: vitamins.0.master, used tablet: qa_iad-1350593002 (172.18.106.144), vttablet: rpc error: code = InvalidArgument desc = Data too long for column \'process\' at row 1 (errno 1406) (sqlstate 22001) (CallerID: acharis.hs-dataops.vitess): Sql: \"insert into foo(id, process) values (:v1, :v2)\", BindVars: {v1: \"type:INT64 value:\\\"1\\\" \"v2: \"type:VARBINARY value:\\\"\\\\312G\\\\307\\\\033n\\\\275L\\\\266\\\\236\\\\004w\\\\002\\\\010\\\\307#G\\\" \"#maxLimit: \"type:INT64 value:\\\"100001\\\" \"}"
code: INVALID_ARGUMENT
 [statement:"INSERT INTO foo (id, process) values (:id, :process)", located:"INSERT INTO foo (id, process) values (:id, :process)", rewritten:"INSERT INTO foo (id, process) values (?, ?)", arguments:{ positional:{}, named:{process:[-54, 71, -57, 27, 110, -67, 76, -74, -98, 4, 119, 2, 8, -57, 35, 71],id:1}, finder:[]}]
        at org.skife.jdbi.v2.SQLStatement.internalExecute(SQLStatement.java:1338)
        at org.skife.jdbi.v2.Update.execute(Update.java:56)
        at org.skife.jdbi.v2.sqlobject.UpdateHandler$2.value(UpdateHandler.java:68)
        at org.skife.jdbi.v2.sqlobject.UpdateHandler.invoke(UpdateHandler.java:81)
        at org.skife.jdbi.v2.sqlobject.SqlObject.invoke(SqlObject.java:216)

```

OLD message returned to java client:
```
org.skife.jdbi.v2.exceptions.UnableToExecuteStatementException: java.sql.SQLNonTransientException: gRPC StatusRuntimeException: hidden.vitess.io.grpc.StatusRuntimeException: CANCELLED: Failed to read message. [statement:"INSERT INTO foo (id, process) values (:id, :process)", located:"INSERT INTO foo (id, process) values (:id, :process)", rewritten:"INSERT INTO foo (id, process) values (?, ?)", arguments:{ positional:{}, named:{process:[73, -17, -47, -84, 18, 50, 70, 6, -81, -35, 1, 78, -121, -113, -39, 94],id:1}, finder:[]}]
        at org.skife.jdbi.v2.SQLStatement.internalExecute(SQLStatement.java:1338)
        at org.skife.jdbi.v2.Update.execute(Update.java:56)
        at org.skife.jdbi.v2.sqlobject.UpdateHandler$2.value(UpdateHandler.java:68)
        at org.skife.jdbi.v2.sqlobject.UpdateHandler.invoke(UpdateHandler.java:81)
        at org.skife.jdbi.v2.sqlobject.SqlObject.invoke(SqlObject.java:216)
        at org.skife.jdbi.v2.sqlobject.SqlObject$2.intercept(SqlObject.java:109)

```

NEW log message:
```
I0413 21:05:10.388922    1748 tabletserver.go:1424] Data too long for column 'process' at row 1 (errno 1406) (sqlstate 22001) (CallerID: acharis.hs-dataops.vitess): Sql: "insert into foo(id, process) values (:v1, :v2)", BindVars: {#maxLimit: "type:INT64 value:\"100001\" "v1: "type:INT64 value:\"1\" "v2: "type:VARBINARY value:\"\\312G\\307\\033n\\275L\\266\\236\\004w\\002\\010\\307#G\" "}

```

OLD log message:
```
I0413 21:06:04.541812     103 tabletserver.go:1391] Data too long for column 'process' at row 1 (errno 1406) (sqlstate 22001) during query: insert into foo(id, process) values (1, 'I�Ѭ2F��N���^'), CallerID: acharis.hs-dataops.vitess: Sql: "insert into foo(id, process) values (:v1, :v2)", BindVars: {v1: "type:INT64 value:\"1\" "v2: "type:VARBINARY value:\"I\\357\\321\\254\\0222F\\006\\257\\335\\001N\\207\\217\\331^\" "#maxLimit: "type:INT64 value:\"100001\" "}

```

acharis commented:
(pulling commit message out into comment)
these are the concerns/constraints i'm trying to balance/satisfy simultaneously:
 - in terse mode, do not put bind variables into error message
 - truncate messages that will appear in tablet logs
 - properly escape values in queries rather than logging/returning non-utf8
 - keep trailing comments
 - testability: map traversal order is randomized, so sort
please let me know if there are others i should keep in mind

acharis commented:
@alainjobart @demmer can we merge this?

demmer commented on go/vt/vtqueryserver/endtoend_test.go:
```diff
@@ -387,10 +387,11 @@ func TestQueryDeadline(t *testing.T) {
-	if err == nil || !strings.Contains(err.Error(), wantErr) {
-		t.Errorf("want error %v, got %v", wantErr, err)
+	wantErr := "EOF (errno 2013) (sqlstate HY000)"
+	if err == nil || !strings.HasPrefix(err.Error(), wantErr) {
```
More curiosity than objection... can this test be changed to do a full equality test on the error instead of just the prefix match?

demmer commented on go/vt/vttablet/tabletserver/tabletserver.go:
```diff
@@ -1402,32 +1391,66 @@ func (tsv *TabletServer) convertAndLogError(ctx context.Context, sql string, bin
-func formatErrorWithCallerID(ctx context.Context, err error) error {
-	if err == nil {
-		return nil
+func doubleTruncate(sql string, bindVariables map[string]*querypb.BindVariable) string {
```
I think renaming to `truncateSqlAndBindVars` would be clearer as would adding a comment.

demmer commented on go/vt/vttablet/endtoend/misc_test.go:
```diff
@@ -564,20 +566,54 @@ func TestDBAStatements(t *testing.T) {
 	}
-	if err.Error() != want {
-		t.Errorf("log was unexpectedly truncated... got %s, wanted %s", err, want)
+	if getTestLog(0) != want {
```
IMO these tests should verify the expected behavior for both the emitted log and the returned error.

sougou commented:
The youtubers seem swamped. Based on the comments, I'm presuming it's safe to merge this. I will wait till tomorrow. If I hear no objections, I'll merge.
